### ISSUE TITLE
Does AFLplusplus compile on macOS?

### ISSUE THREAD

**Issue by prRZ5F4LXZ** [AFLplusplus/AFLplusplus#858]
**IMPORTANT**
1. You have verified that the issue to be present in the current `dev` branch

I just used the clone as I don't think the problem is specific to dev.

2. Please supply the command line options and relevant environment variables, e.g. a copy-paste of the contents of `out/default/fuzzer_setup`

LLVM is in the PATH

/usr/local/opt/llvm/bin

**Describe the bug**

I run the following command in macOS Catalina. But it fails.

```
$ make distrib
[*] Checking for the default compiler cc...
[*] Testing the PATH environment variable...
[*] Checking for the ability to compile x86 code...
[+] shmat seems to be working.
[+] Python 3.7.6 support seems to be working.
[+] Everything seems to be working, ready to compile.
make -j -f GNUmakefile.llvm
make[1]: Entering directory '/private/tmp/AFLplusplus'
[+] llvm_mode detected llvm 10+, enabling neverZero implementation and c++14
[+] llvm_mode detected llvm 11+, enabling afl-lto LTO implementation
[+] shmat seems to be working.
[*] Checking for working 'llvm-config'...
[*] Checking for working '/usr/local/Cellar/llvm/11.0.0_1/bin/clang'...
[*] Checking for matching versions of '/usr/local/Cellar/llvm/11.0.0_1/bin/clang' and 'llvm-config'
[*] We have llvm-config version 11.0.0 with a clang version 11.0.0, good.
[*] Checking for './afl-showmap'...
[+] All set and ready to build.
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -Wno-cast-qual -Wno-variadic-macros -Wno-pointer-sign -I ./include/ -I ./instrumentation/ -DAFL_PATH=\"/usr/local/lib/afl\" -DBIN_PATH=\"/usr/local/bin\" -DLLVM_BINDIR=\"/usr/local/Cellar/llvm/11.0.0_1/bin\" -DVERSION=\"++3.12c\" -DLLVM_LIBDIR=\"/usr/local/Cellar/llvm/11.0.0_1/lib\" -DLLVM_VERSION=\"11.0.0\" -Wno-deprecated -DAFL_CLANG_FLTO=\"-flto=full\" -DAFL_REAL_LD=\"/usr/local/Cellar/llvm/11.0.0_1/bin/ld.lld\" -DAFL_CLANG_LDPATH=\"\" -DAFL_CLANG_FUSELD=\"1\" -DCLANG_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang\" -DCLANGPP_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang++\" -DUSE_BINDIR=1 -Wno-unused-function -fdebug-prefix-map="/private/tmp/AFLplusplus=llvm_mode"  `llvm-config --cxxflags` -fno-rtti -fPIC -std=c++14 -c instrumentation/afl-llvm-common.cc -o instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -DLLVMInsTrim_EXPORTS -fno-rtti -fPIC -std=c++14 -shared instrumentation/afl-llvm-pass.so.cc -o afl-llvm-pass.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -fno-rtti -fPIC -std=c++14 -shared instrumentation/SanitizerCoveragePCGUARD.so.cc -o SanitizerCoveragePCGUARD.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -shared instrumentation/split-compares-pass.so.cc -o split-compares-pass.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -shared instrumentation/split-switches-pass.so.cc -o split-switches-pass.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -shared instrumentation/cmplog-routines-pass.cc -o cmplog-routines-pass.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -shared instrumentation/cmplog-instructions-pass.cc -o cmplog-instructions-pass.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -shared instrumentation/afl-llvm-dict2file.so.cc -o afl-llvm-dict2file.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -shared instrumentation/compare-transform-pass.so.cc -o compare-transform-pass.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -DLLVMInsTrim_EXPORTS -fno-rtti -fPIC -std=c++14 -shared instrumentation/LLVMInsTrim.so.cc instrumentation/MarkNodes.cc -o libLLVMInsTrim.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -Wno-cast-qual -Wno-variadic-macros -Wno-pointer-sign -I ./include/ -I ./instrumentation/ -DAFL_PATH=\"/usr/local/lib/afl\" -DBIN_PATH=\"/usr/local/bin\" -DLLVM_BINDIR=\"/usr/local/Cellar/llvm/11.0.0_1/bin\" -DVERSION=\"++3.12c\" -DLLVM_LIBDIR=\"/usr/local/Cellar/llvm/11.0.0_1/lib\" -DLLVM_VERSION=\"11.0.0\" -Wno-deprecated -DAFL_CLANG_FLTO=\"-flto=full\" -DAFL_REAL_LD=\"/usr/local/Cellar/llvm/11.0.0_1/bin/ld.lld\" -DAFL_CLANG_LDPATH=\"\" -DAFL_CLANG_FUSELD=\"1\" -DCLANG_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang\" -DCLANGPP_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang++\" -DUSE_BINDIR=1 -Wno-unused-function -fdebug-prefix-map="/private/tmp/AFLplusplus=llvm_mode"  src/afl-ld-lto.c -o afl-ld-lto
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -fno-rtti -fPIC -std=c++14 -shared instrumentation/afl-llvm-lto-instrumentlist.so.cc -o afl-llvm-lto-instrumentlist.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -Wno-writable-strings -fno-rtti -fPIC -std=c++14 -shared instrumentation/afl-llvm-lto-instrumentation.so.cc -o afl-llvm-lto-instrumentation.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
instrumentation/afl-llvm-lto-instrumentation.so.cc:179:58: warning: format specifies type 'unsigned long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
  if (debug) { fprintf(stderr, "map address is 0x%lx\n", map_addr); }
                                                 ~~~     ^~~~~~~~
                                                 %llx
1 warning generated.
/usr/local/Cellar/llvm/11.0.0_1/bin/clang -Wall -g -Wno-cast-qual -Wno-variadic-macros -Wno-pointer-sign -I ./include/ -I ./instrumentation/ -DAFL_PATH=\"/usr/local/lib/afl\" -DBIN_PATH=\"/usr/local/bin\" -DLLVM_BINDIR=\"/usr/local/Cellar/llvm/11.0.0_1/bin\" -DVERSION=\"++3.12c\" -DLLVM_LIBDIR=\"/usr/local/Cellar/llvm/11.0.0_1/lib\" -DLLVM_VERSION=\"11.0.0\" -Wno-deprecated -DAFL_CLANG_FLTO=\"-flto=full\" -DAFL_REAL_LD=\"/usr/local/Cellar/llvm/11.0.0_1/bin/ld.lld\" -DAFL_CLANG_LDPATH=\"\" -DAFL_CLANG_FUSELD=\"1\" -DCLANG_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang\" -DCLANGPP_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang++\" -DUSE_BINDIR=1 -Wno-unused-function -fdebug-prefix-map="/private/tmp/AFLplusplus=llvm_mode"  -Wno-unused-result -O0 -flto=full -fPIC -c instrumentation/afl-llvm-rt-lto.o.c -o ./afl-llvm-rt-lto.o
/usr/local/Cellar/llvm/11.0.0_1/bin/clang++ `llvm-config --cxxflags` -fno-rtti -fPIC -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -I ./include/ -DVERSION=\"++3.12c\" -Wno-variadic-macros -DLLVM_MINOR=0 -DLLVM_MAJOR=11 -Wno-deprecated-declarations -Wno-writable-strings -fno-rtti -fPIC -std=c++14 -shared instrumentation/SanitizerCoverageLTO.so.cc -o SanitizerCoverageLTO.so `llvm-config --ldflags`  -Wl,-flat_namespace -Wl,-undefined,suppress instrumentation/afl-llvm-common.o
[*] Testing the CC wrapper and instrumentation output...
unset AFL_USE_ASAN AFL_USE_MSAN AFL_INST_RATIO; ASAN_OPTIONS=detect_leaks=0 AFL_QUIET=1 AFL_PATH=. AFL_LLVM_LAF_ALL=1 ./afl-cc -O3 -funroll-loops -fPIC -D_FORTIFY_SOURCE=2 -Wall -g -Wno-cast-qual -Wno-variadic-macros -Wno-pointer-sign -I ./include/ -I ./instrumentation/ -DAFL_PATH=\"/usr/local/lib/afl\" -DBIN_PATH=\"/usr/local/bin\" -DLLVM_BINDIR=\"/usr/local/Cellar/llvm/11.0.0_1/bin\" -DVERSION=\"++3.12c\" -DLLVM_LIBDIR=\"/usr/local/Cellar/llvm/11.0.0_1/lib\" -DLLVM_VERSION=\"11.0.0\" -Wno-deprecated -DAFL_CLANG_FLTO=\"-flto=full\" -DAFL_REAL_LD=\"/usr/local/Cellar/llvm/11.0.0_1/bin/ld.lld\" -DAFL_CLANG_LDPATH=\"\" -DAFL_CLANG_FUSELD=\"1\" -DCLANG_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang\" -DCLANGPP_BIN=\"/usr/local/Cellar/llvm/11.0.0_1/bin/clang++\" -DUSE_BINDIR=1 -Wno-unused-function -fdebug-prefix-map="/private/tmp/AFLplusplus=llvm_mode"  ./test-instr.c -o test-instr
ASAN_OPTIONS=detect_leaks=0 ./afl-showmap -m none -q -o .test-instr0 ./test-instr < /dev/null

[-]  SYSTEM ERROR : shmget() failed
    Stop location : afl_shm_init(), src/afl-sharedmem.c:247
       OS message : Invalid argument
make[1]: *** [GNUmakefile.llvm:461: test_build] Error 1
make[1]: Leaving directory '/private/tmp/AFLplusplus'
make: [GNUmakefile:308: llvm] Error 2 (ignored)
make -f GNUmakefile.gcc_plugin
make[1]: Entering directory '/private/tmp/AFLplusplus'
[+] shmat seems to be working.
[*] Checking for working 'gcc'...
[*] Checking for gcc plugin development header files...
[-] Oops, can't find gcc header files. Be sure to install 'gcc-X-plugin-dev'.
make[1]: *** [GNUmakefile.gcc_plugin:121: test_deps] Error 1
make[1]: Leaving directory '/private/tmp/AFLplusplus'
make: [GNUmakefile:313: gcc_plugin] Error 2 (ignored)
[*] Testing the CC wrapper afl-cc and its instrumentation output...
afl-cc ++3.12c by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: GCC-GCC
[!] WARNING: You are using outdated instrumentation, install LLVM and/or gcc-plugin and use afl-clang-fast/afl-clang-lto/afl-gcc-fast instead!
ASAN_OPTIONS=detect_leaks=0 ./afl-showmap -m none -q -o .test-instr0 ./test-instr < /dev/null

[-]  SYSTEM ERROR : shmget() failed
    Stop location : afl_shm_init(), src/afl-sharedmem.c:247
       OS message : Invalid argument
make: *** [GNUmakefile:521: test_build] Error 1
```

**To Reproduce**

See above.

**Expected behavior**

Compile the source successfully.

**Screen output/Screenshots**

See above.

**Additional context**

NA

**Comment by vanhauser-thc** [comment]
have you increased the shared memory size?
MacOS has a ridiculous small default size

**Comment by prRZ5F4LXZ** [comment]
Could you show me the exact commands on how to increase it on Catalina? I am not familiar with shared memory conf on macOS.

**Comment by vanhauser-thc** [comment]
I do not have a Mac. I am pretty sure that information is easy to find with Google, did that or a friend like a year ago.

**Comment by prRZ5F4LXZ** [comment]
Maybe somebody can add this info to the installation instruction?

**Comment by vanhauser-thc** [comment]
fuzzing is a lot of hands one and troubleshooting. if you fail this step alreay I would recommend to build more basic knowledge about systems, developing and building first.
feel free to submit a PR to add this information.

**Comment by domenukk** [comment]
For MacOS: take a look at https://serverfault.com/questions/10226/how-to-set-the-shmmax-parameter-on-mac-os-x
also, `ipcs` to list and `ipcrm` to remove shared maps are your friend.

For cleaning, see https://stackoverflow.com/questions/8150562/cleaning-semaphores-in-mac-os-x

**Comment by domenukk** [comment]
TL;DR: `ipcs -m |  cut -f2 -d' ' | xargs -I {} sudo ipcrm -m {}; ipcs -m |  cut -f3 -d' ' | xargs -I {} sudo ipcrm -m {}` to remove old maps, then
```bash
sudo sysctl -w kern.sysv.shmall=65536
sudo sysctl -w kern.sysv.shmmax=16777216
```
(or some other high number) to increase count and size.
